src/core/Akka/Serialization/CompatJsonSerializer.cs

```diff
@@ -0,0 +1,302 @@
+                    : PreserveReferencesHandling.None,
+                Converters = converters,
+                NullValueHandling = NullValueHandling.Ignore,
+                DefaultValueHandling = DefaultValueHandling.Ignore,
+                MissingMemberHandling = MissingMemberHandling.Ignore,
+                ObjectCreationHandling = ObjectCreationHandling.Replace, //important: if reuse, the serializer will overwrite properties in default references, e.g. Props.DefaultDeploy or Props.noArgs
+                ConstructorHandling = ConstructorHandling.AllowNonPublicDefaultConstructor,
+                TypeNameHandling = settings.EncodeTypeNames
+                    ? TypeNameHandling.All
+                    : TypeNameHandling.None,
+                ContractResolver = new AkkaContractResolver()
+            };
+
+            _serializer = JsonSerializer.Create(Settings);
+        }
+
+        private static JsonConverter CreateConverter(Type converterType, ExtendedActorSystem actorSystem)
+        {
+            var ctor = converterType.GetConstructors()
+                .FirstOrDefault(c =>
+                {
+                    var parameters = c.GetParameters();
+                    return parameters.Length == 1 && parameters[0].ParameterType == typeof(ExtendedActorSystem);
+                });
+
+            return ctor == null
+                ? (JsonConverter)Activator.CreateInstance(converterType)
+                : (JsonConverter)Activator.CreateInstance(converterType, actorSystem);
+        }
+
+        internal class AkkaContractResolver : DefaultContractResolver
+        {
+            protected override JsonProperty CreateProperty(MemberInfo member, MemberSerialization memberSerialization)
+            {
+                var prop = base.CreateProperty(member, memberSerialization);
+
+                if (!prop.Writable)
+                {
+                    var property = member as PropertyInfo;
+                    if (property != null)
+                    {
+                        var hasPrivateSetter = property.GetSetMethod(true) != null;
+                        prop.Writable = hasPrivateSetter;
+                    }
+                }
+
+                return prop;
+            }
+        }
+
+        /// <inheritdoc />
+        /// <summary>
+        /// Returns whether this serializer needs a manifest in the fromBinary method
+        /// </summary>
+        public override bool IncludeManifest => false;
+
+        /// <inheritdoc />
+        /// <summary>
+        /// Serializes the given object into a byte array
+        /// </summary>
+        /// <param name="obj">The object to serialize </param>
+        /// <returns>A byte array containing the serialized object</returns>
+        public override byte[] ToBinary(object obj)
+        {
+            string data = JsonConvert.SerializeObject(obj, Formatting.None, Settings);
+#if CORECLR
+            data = data.Replace("System.Private.CoreLib", "%CORE%");
```

Comment:
it is not enough. There is more `System.Private.*` assemblies, like `System.Private.Uri`